### ISSUE TITLE
Kubelet does not retry init container startup on CNI failure

### ISSUE THREAD

**Issue by erhudy** [kubernetes/kubernetes#78125]
**What happened**: A user reported that a workload was stuck and unable to startup. The workload in question spawns several pods via a CRD, one of which is an init container that checks the other pods for readiness before the workload begins processing. During startup, the CNI call failed due to a timeout communicating with the API server due to a transient network issue. Instead of retrying, kubelet goes into a state where it ignores the pod in question until kubelet is restarted, at which point it notices the unstarted init container and runs it.

**What you expected to happen**: Kubelet should report the init container as having failed, triggering normal backoff restartPolicy.

**How to reproduce it (as minimally and precisely as possible)**: We don't have a consistent reproducer but launching jobs while fuzzing the API server seems like it would probably be able to trigger the failure.

**Anything else we need to know?**: Not that I know of.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.14.0
- Cloud provider or hardware configuration: bare metal
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04
- Kernel (e.g. `uname -a`): 4.8.0-41-generic
- Install tools: internal tooling
- Network plugin and version (if this is a network-related bug): Calico 3.6.1
- Others:

**Comment by erhudy** [comment]
See sanitized log below:
```
May 20 09:05:41 worker-001 kubelet[152698]: I0520 09:05:41.638702  152698 reconciler.go:207] operationExecutor.VerifyControllerAttachedVolume started for volume "default-token-wb5gm" (UniqueName: "kubernetes.io/secret/715e8cb1-7ade-11e9-8520-000f533064c0-default-token-wb5gm") pod "workload-f85f9-driver" (UID: "715e8cb1-7ade-11e9-8520-000f533064c0")
May 20 09:05:41 worker-001 kubelet[152698]: I0520 09:05:41.638783  152698 reconciler.go:207] operationExecutor.VerifyControllerAttachedVolume started for volume "job-data" (UniqueName: "kubernetes.io/empty-dir/715e8cb1-7ade-11e9-8520-000f533064c0-job-data") pod "workload-f85f9-driver" (UID: "715e8cb1-7ade-11e9-8520-000f533064c0")
May 20 09:06:16 worker-001 kubelet[152698]: E0520 09:06:16.332878  152698 cni.go:331] Error adding userns1_workload-f85f9-driver/1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38 to network calico/k8s-pod-network: Get https://[10.96.0.1]:443/api/v1/namespaces/userns1: net/http: TLS handshake timeout
May 20 09:06:23 worker-001 kubelet[152698]: 2019-05-20 09:06:23.547 [WARNING][58045] k8s.go 447: WorkloadEndpoint does not exist in the datastore, moving forward with the clean up ContainerID="1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" WorkloadEndpoint="worker--001-k8s-workload--f85f9--driver-eth0"
May 20 09:06:39 worker-001 kubelet[152698]: 2019-05-20 09:06:39.348 [WARNING][58103] ipam_plugin.go 306: Asked to release address but it doesn't exist. Ignoring ContainerID="1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" HandleID="k8s-pod-network.1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" Workload="worker--001-k8s-workload--f85f9--driver-eth0"
May 20 09:06:40 worker-001 kubelet[152698]: E0520 09:06:40.462362  152698 remote_runtime.go:109] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = failed to set up sandbox container "1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" network for pod "workload-f85f9-driver": NetworkPlugin cni failed to set up pod "workload-f85f9-driver_userns1" network: Get https://[10.96.0.1]:443/api/v1/namespaces/userns1: net/http: TLS handshake timeout
May 20 09:06:40 worker-001 kubelet[152698]: E0520 09:06:40.462468  152698 kuberuntime_sandbox.go:68] CreatePodSandbox for pod "workload-f85f9-driver_userns1(715e8cb1-7ade-11e9-8520-000f533064c0)" failed: rpc error: code = Unknown desc = failed to set up sandbox container "1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" network for pod "workload-f85f9-driver": NetworkPlugin cni failed to set up pod "workload-f85f9-driver_userns1" network: Get https://[10.96.0.1]:443/api/v1/namespaces/userns1: net/http: TLS handshake timeout
May 20 09:06:40 worker-001 kubelet[152698]: E0520 09:06:40.462508  152698 kuberuntime_manager.go:693] createPodSandbox for pod "workload-f85f9-driver_userns1(715e8cb1-7ade-11e9-8520-000f533064c0)" failed: rpc error: code = Unknown desc = failed to set up sandbox container "1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" network for pod "workload-f85f9-driver": NetworkPlugin cni failed to set up pod "workload-f85f9-driver_userns1" network: Get https://[10.96.0.1]:443/api/v1/namespaces/userns1: net/http: TLS handshake timeout
May 20 09:06:40 worker-001 kubelet[152698]: E0520 09:06:40.462598  152698 pod_workers.go:190] Error syncing pod 715e8cb1-7ade-11e9-8520-000f533064c0 ("workload-f85f9-driver_userns1(715e8cb1-7ade-11e9-8520-000f533064c0)"), skipping: failed to "CreatePodSandbox" for "workload-f85f9-driver_userns1(715e8cb1-7ade-11e9-8520-000f533064c0)" with CreatePodSandboxError: "CreatePodSandbox for pod \"workload-f85f9-driver_userns1(715e8cb1-7ade-11e9-8520-000f533064c0)\" failed: rpc error: code = Unknown desc = failed to set up sandbox container \"1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38\" network for pod \"workload-f85f9-driver\": NetworkPlugin cni failed to set up pod \"workload-f85f9-driver_userns1\" network: Get https://[10.96.0.1]:443/api/v1/namespaces/userns1: net/http: TLS handshake timeout"
May 20 09:06:59 worker-001 kubelet[152698]: W0520 09:06:59.348953  152698 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "workload-f85f9-driver_userns1": CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38"
May 20 09:07:08 worker-001 kubelet[152698]: 2019-05-20 09:07:08.387 [WARNING][58449] k8s.go 447: WorkloadEndpoint does not exist in the datastore, moving forward with the clean up ContainerID="1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" WorkloadEndpoint="worker--001-k8s-workload--f85f9--driver-eth0"
May 20 09:07:16 worker-001 kubelet[152698]: 2019-05-20 09:07:16.636 [WARNING][58509] ipam_plugin.go 306: Asked to release address but it doesn't exist. Ignoring ContainerID="1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" HandleID="k8s-pod-network.1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" Workload="worker--001-k8s-workload--f85f9--driver-eth0"

[...I restart Kubelet on the node...]

May 20 15:46:36 worker-001 kubelet[162001]: 2019-05-20 15:46:36.254 [WARNING][162150] k8s.go 447: WorkloadEndpoint does not exist in the datastore, moving forward with the clean up ContainerID="1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" WorkloadEndpoint="worker--001-k8s-workload--f85f9--driver-eth0"
May 20 15:46:36 worker-001 kubelet[162001]: 2019-05-20 15:46:36.294 [WARNING][162168] ipam_plugin.go 306: Asked to release address but it doesn't exist. Ignoring ContainerID="1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" HandleID="k8s-pod-network.1fcc27225ffd7583ff1d13ac5f70f4ab3b82a9651e26b93d21c5afa72d660f38" Workload="worker--001-k8s-workload--f85f9--driver-eth0"
May 20 15:46:52 worker-001 kubelet[162001]: I0520 15:46:52.464095  162001 reconciler.go:207] operationExecutor.VerifyControllerAttachedVolume started for volume "job-data" (UniqueName: "kubernetes.io/empty-dir/715e8cb1-7ade-11e9-8520-000f533064c0-job-data") pod "workload-f85f9-driver" (UID: "715e8cb1-7ade-11e9-8520-000f533064c0")
May 20 15:46:52 worker-001 kubelet[162001]: I0520 15:46:52.465404  162001 reconciler.go:207] operationExecutor.VerifyControllerAttachedVolume started for volume "default-token-wb5gm" (UniqueName: "kubernetes.io/secret/715e8cb1-7ade-11e9-8520-000f533064c0-default-token-wb5gm") pod "workload-f85f9-driver" (UID: "715e8cb1-7ade-11e9-8520-000f533064c0")
May 20 15:46:53 worker-001 kubelet[162001]: E0520 15:46:53.606844  162001 nestedpendingoperations.go:267] Operation for "\"kubernetes.io/secret/715e8cb1-7ade-11e9-8520-000f533064c0-default-token-wb5gm\" (\"715e8cb1-7ade-11e9-8520-000f533064c0\")" failed. No retries permitted until 2019-05-20 15:46:54.106775976 +0000 UTC m=+18.475745410 (durationBeforeRetry 500ms). Error: "MountVolume.SetUp failed for volume \"default-token-wb5gm\" (UniqueName: \"kubernetes.io/secret/715e8cb1-7ade-11e9-8520-000f533064c0-default-token-wb5gm\") pod \"workload-f85f9-driver\" (UID: \"715e8cb1-7ade-11e9-8520-000f533064c0\") : couldn't propagate object cache: timed out waiting for the condition"

[...pod starts up and runs normally...]
```

**Comment by svenbs** [comment]
This also happens in our environment for cronjobs quite regularly.
Because of different reasons cronjob pods sandboxes wont be created. Resulting in hanging cronjobs getting overdued. This also triggers #42649 for us (jfi).

**Comment by johnmarcou** [comment]
Same here, K8s 1.14.2.

@svenbs, is that happen to you on initContainers as well? Or just container?

**Comment by svenbs** [comment]
@johnmarcou: It's happening for us while the container sandbox is created. The pod gets stuck in the state "ContainerCreating".

I would assume this happens on both initcontainers and containers.

**Comment by johnmarcou** [comment]
Probably closed by #79451
